### ISSUE TITLE
[Bug] Samples generated using draw_sobol_samples do not span the sample space

### ISSUE THREAD

**Issue by saitcakmak** [pytorch/botorch#504]
# 🐛 Bug

When generating samples using `draw_sobol_samples`, the resulting samples do not span the solution space for each q-batch. The example below shows that when drawing samples from 1 dimensional sample space, some of the q-batches only have samples from [0, 0.5]. This appears to be a result of `draw_sobol_samples` reshaping the `(n * q) x d` samples generated using the `SobolEngine` into `n x q x d`.

## To reproduce

** Code snippet to reproduce **
```python
import torch
from botorch.utils import draw_sobol_samples

samples = draw_sobol_samples(
    bounds=torch.tensor([[0.], [1.]]), n=4000, q=100
)
max, _ = torch.max(samples, dim=0)
print(max)

```

** Stack trace/error message **
Max above will include some entries as `0.4999` and `0.5000`, meaning that all samples corresponding to those q-batches come from [0, 0.5].

## Expected Behavior

All entries of max should be very close to 1, meaning that the samples span the sample space [0, 1] for each q-batch.

## System information

**Please complete the following information:**
BoTorch 0.3.0
GPyTorch 1.1.1
PyTorch 1.5.1

**Comment by Balandat** [comment]
Thanks for flagging, very interesting.

Some digging shows that the samples are uniform in all `q`-batches as long as `q<100`, for `q=100` this breaks and we get the degenerate behavior that you reported. This suggests this is an issue with the `SobolEngine` in torch. I'll take a look upstream to see what's going on.

I guess we haven't really tested this rigorously with `q`-batches that large, so that's why we haven't run into this issue before.

**Comment by Balandat** [comment]
FWIW as another pointer, replacing  `.view(n, q, d)` in `draw_sobol_samples` with `.view(q, n, d).permute(1, 0, 2)` seems to avoid this issue. Still not sure if a problem with the qMC logic or some issue with strides or the like.

**Comment by saitcakmak** [comment]
I wonder if that would also break for some choices of n and q. 

My guess is that since the sobol samples have an inherent structure (generated by modifying uniform partitions), they're not suitable for reshaping in general. We can project the low discrepancy samples to a lower dimension and keep the nice properties, but we can't, in general, combine low dimensional samples to generate large dimensional low discrepancy samples. Reshaping `(n * q) x d` samples into `n x q x d` is in some sense using `d` dimensional samples to generate larger `q * d` dimensional samples. This might be why it breaks under certain choices of n and q.

This is based on a discussion with @peter-i-frazier and @RaulAstudillo06. Tagging in case they have something to add.

**Comment by Balandat** [comment]
That much is understood. I guess it's really just a simple bug then - instead of 
```
sobol_engine = SobolEngine(d, scramble=True, seed=seed)
samples_raw = sobol_engine.draw(n * q, dtype=lower.dtype).view(n, q, d)
```
this should be 
```
sobol_engine = SobolEngine(q * d, scramble=True, seed=seed)
samples_raw = sobol_engine.draw(n, dtype=lower.dtype).view(n, q, d)
```
as this is the effective dimension. 

Note that this is also the implementation in `SobolQMCNormalSampler`, so we should be ok there: https://github.com/pytorch/botorch/blob/master/botorch/sampling/samplers.py#L284-L290

I'll put up a PR for this one.

**Comment by Balandat** [comment]
FWIW I have an upstream issue for updating the direction numbers of the Sobol sampler: https://github.com/pytorch/pytorch/issues/41489, which will allow higher effective dimensions `q * d` soon.

**Comment by saitcakmak** [comment]
Yeah, that's a great quick fix. With the new direction numbers, it shouldn't cause any issues even with reasonably large `q`. 

FWIW, here is another issue I noticed with sobol sampling: Unless you specify the seed when calling `draw_sobol_samples(..., seed=seed)`, `SobolEngine` calls `torch.Generator.seed()` and sets a completely random seed that is independent of the random number generator that is used for the rest of the code. Not a big deal for general use, but it causes some issues with reproducibility as it requires you to pass a seed to each call to `draw_sobol_samples`, and just setting `torch.manual_seed(seed)` in the beginning of the code doesn't produce reproducible results. Just wanted to bring it up while we were on the issue. If you think it is worth changing, a simple fix would be to generate `seed = torch.randint(some_large_number, (1,))` in `draw_sobol_samples` and to pass this to `SobolEngine`.

**Comment by Balandat** [comment]
Actually, I should have fixed this upstream in https://github.com/pytorch/pytorch/pull/36427. This makes the behavior consistent with the other random functions in that it uses the default RNG if seed is not specified. This should be in a recent pytorch version, certainly in the most recent 1.6.

**Comment by saitcakmak** [comment]
Oh, cool! I will update to the most recent version. Thanks for pointing that out!

**Comment by Balandat** [comment]
I also put a cython version of the sobol generator with the new direction numbers into scipy in case you want to play around with that before I get to the torch update: https://github.com/scipy/scipy/pull/10844

**Comment by saitcakmak** [comment]
Cool, I'll check that out
